### PR TITLE
Allow longer timeout for starting pageserver, safe keeper and storage controller in test cases to make test cases less flaky

### PR DESCRIPTION
**Pull request by Bodobolero**
## Problem

see https://github.com/neondatabase/neon/issues/8070

## Summary of changes

the neon_local subcommands to 
- start neon
- start pageserver
- start safekeeper
- start storage controller

get a new option -t=xx  or --start-timeout=xx which allows to specify a longer timeout in seconds we wait for the process start.
This is useful in test cases where the pageserver has to read a lot of layer data, like in pagebench test cases.

In addition we exploit the new timeout option in the python test infrastructure (python fixtures) and modify the flaky testcase to increase the timeout from 10 seconds to 1 minute. 

Example from the test execution

```bash
RUST_BACKTRACE=1 NEON_ENV_BUILDER_USE_OVERLAYFS_FOR_SNAPSHOTS=1 DEFAULT_PG_VERSION=15 BUILD_TYPE=release     ./scripts/pytest test_runner/performance/pageserver/pagebench/test_pageserver_max_throughput_getpage_at_latest_lsn.py
...
2024-06-19 09:29:34.590 INFO [neon_fixtures.py:1513] Running command "/instance_store/neon/target/release/neon_local storage_controller start --start-timeout=60s"
2024-06-19 09:29:36.365 INFO [broker.py:34] starting storage_broker to listen incoming connections at "127.0.0.1:15001"
2024-06-19 09:29:36.365 INFO [neon_fixtures.py:1513] Running command "/instance_store/neon/target/release/neon_local pageserver start --id=1 --start-timeout=60s"
2024-06-19 09:29:36.366 INFO [neon_fixtures.py:1513] Running command "/instance_store/neon/target/release/neon_local safekeeper start 1 --start-timeout=60s"
```

### CODE REVIEW

File: logfile
Comment (bayandin) [inline logfile:1]: We probably don't need to include this file 😄

File: control_plane/src/bin/neon_local.rs [hunk 1]
```diff
@@ -1288,6 +1297,14 @@ async fn try_stop_all(env: &local_env::LocalEnv, immediate: bool) {
 }
 
 fn cli() -> Command {
+    let timeout_arg = Arg::new("timeout-in-seconds")
+        .long("t")
```
Comment (bayandin) [inline control_plane/src/bin/neon_local.rs:1301]: Maybe `--timeout` and `-t`?

Comment (bayandin) [reply control_plane/src/bin/neon_local.rs:1301]: Or ever `--start-timeout`

File: logfile
Comment (Bodobolero) [reply logfile:1]: sorry

File: control_plane/src/bin/neon_local.rs [hunk 2]
```diff
@@ -1202,15 +1220,18 @@ async fn handle_safekeeper(sub_match: &ArgMatches, env: &local_env::LocalEnv) ->
     Ok(())
 }
 
-async fn handle_start_all(env: &local_env::LocalEnv) -> anyhow::Result<()> {
+async fn handle_start_all(
+    env: &local_env::LocalEnv,
+    retry_timeout_in_seconds: u64,
```
Comment (problame) [inline control_plane/src/bin/neon_local.rs:1225]: Why not  `std::time::Duration`?

With that, you can shorted the variable name to `retry_timeout` (please propagate that change down the call graph) and there's more type safety.

**Comment by hlinnaka** [comment]
No objections, but I wonder if this is just papering over the real issue:

>  This is useful in test cases where the pageserver has to read a lot of layer data, like in pagebench test cases.

Why does having a lot of layer data make the pageserver start up slow? There shouldn't be anything in the startup codepath that scales with the amount of data or # of layers, right?

File: control_plane/src/bin/neon_local.rs [hunk 3]
```diff
@@ -1288,6 +1309,15 @@ async fn try_stop_all(env: &local_env::LocalEnv, immediate: bool) {
...
 
 fn cli() -> Command {
+    let timeout_arg = Arg::new("timeout-in-seconds")
+        .long("start-timeout")
+        .short('t')
+        .global(true)
+        .help("timeout in seconds until we fail the command")
+        .value_parser(value_parser!(u64))
+        .default_value("10")
+        .required(false);
```
Comment (problame) [inline control_plane/src/bin/neon_local.rs:1319]: Try parsing a Duration here. `humantime::Duration` implements the `FromStr` trait


```suggestion
        .help("timeout in seconds until we fail the command")
        .value_parser(value_parser!(humantime::Duration))
        .default_value("10s")
        .required(false);
```

File: control_plane/src/bin/neon_local.rs [hunk 4]
```diff
@@ -49,6 +49,7 @@ use utils::{
 const DEFAULT_SAFEKEEPER_ID: NodeId = NodeId(1);
 const DEFAULT_PAGESERVER_ID: NodeId = NodeId(1);
 const DEFAULT_BRANCH_NAME: &str = "main";
+const DEFAULT_RETRY_UNTIL_SECS: u64 = 10;
```
Comment (problame) [inline control_plane/src/bin/neon_local.rs:52]: WHy is this needed? We have the `default_value("10")` for the CLI flag.

File: control_plane/src/background_process.rs [hunk 5]
```diff
@@ -60,6 +61,7 @@ pub async fn start_process<F, Fut, AI, A, EI>(
     envs: EI,
     initial_pid_file: InitialPidFile,
     process_status_check: F,
+    retry_timeout_in_seconds: Option<u64>,
 ) -> anyhow::Result<()>
```
Comment (problame) [inline control_plane/src/background_process.rs:65]: 1. Put the arg before the `process_status_check` closure. Call sites are easier to read that way (closure comes last).

2. All call sites use `Some()`, no need for an `Option<>`.

File: control_plane/src/background_process.rs [hunk 6]
```diff
@@ -36,8 +36,8 @@ use utils::pid_file::{self, PidFileRead};
 // it's waiting. If the process hasn't started/stopped after 5 seconds,
 // it prints a notice that it's taking long, but keeps waiting.
 //
-const RETRY_UNTIL_SECS: u64 = 10;
-const RETRIES: u64 = (RETRY_UNTIL_SECS * 1000) / RETRY_INTERVAL_MILLIS;
+const DEFAULT_RETRY_TIMEOUT_SECS: u64 = 10;
```
Comment (problame) [inline control_plane/src/background_process.rs:39]: This seems like yet another duplicate of the default 10s.

**Review by problame — CHANGES_REQUESTED** [review]

File: control_plane/src/bin/neon_local.rs (hunk 2 above)
Comment (Bodobolero) [reply control_plane/src/bin/neon_local.rs:1225]: good point, the const variable used before for the retry timeout was a u64 and I "sticked to the convention" but you are right

**Comment by Bodobolero** [comment]
> No objections, but I wonder if this is just papering over the real issue:
> 
> > This is useful in test cases where the pageserver has to read a lot of layer data, like in pagebench test cases.
> 
> Why does having a lot of layer data make the pageserver start up slow? There shouldn't be anything in the startup codepath that scales with the amount of data or # of layers, right?

We discussed this on slack and agreed on increasing the timeout

File: control_plane/src/background_process.rs (hunk 6 above)
Comment (Bodobolero) [reply control_plane/src/background_process.rs:39]: This is a default for stopping the servers

File: control_plane/src/background_process.rs (hunk 5 above)
Comment (Bodobolero) [reply control_plane/src/background_process.rs:65]: done

File: control_plane/src/bin/neon_local.rs (hunk 4 above)
Comment (Bodobolero) [reply control_plane/src/bin/neon_local.rs:52]: removed it

File: control_plane/src/bin/neon_local.rs (hunk 3 above)
Comment (Bodobolero) [reply control_plane/src/bin/neon_local.rs:1319]: done

File: control_plane/src/bin/neon_local.rs [hunk 7]
```diff
@@ -1046,10 +1047,20 @@ fn get_pageserver(env: &local_env::LocalEnv, args: &ArgMatches) -> Result<PageSe
     ))
 }
 
+fn get_retry_timeout(args: &ArgMatches) -> &Duration {
```
Comment (problame) [inline control_plane/src/bin/neon_local.rs:1050]: nit: let's call the method like the flag & variable, i.e., `get_start_timeout`

File: control_plane/src/background_process.rs [hunk 8]
```diff
@@ -36,11 +36,12 @@ use utils::pid_file::{self, PidFileRead};
 // it's waiting. If the process hasn't started/stopped after 5 seconds,
 // it prints a notice that it's taking long, but keeps waiting.
 //
-const RETRY_UNTIL_SECS: u64 = 10;
-const RETRIES: u64 = (RETRY_UNTIL_SECS * 1000) / RETRY_INTERVAL_MILLIS;
-const RETRY_INTERVAL_MILLIS: u64 = 100;
-const DOT_EVERY_RETRIES: u64 = 10;
-const NOTICE_AFTER_RETRIES: u64 = 50;
+const DEFAULT_STOP_RETRY_TIMEOUT: Duration = Duration::from_secs(10);
+const DEFAULT_STOP_RETRIES: u128 =
```
Comment (problame) [inline control_plane/src/background_process.rs:40]: nit: (possibly I suggested the wrong thing earlier): there is no way to change stop retry count or timeout, so, the `DEFAULT_` prefix is redundant / misleading.

**Review by problame — APPROVED** [review]
+1 for rust bits

File: control_plane/src/bin/neon_local.rs (hunk 7 above)
Comment (Bodobolero) [reply control_plane/src/bin/neon_local.rs:1050]: ok

**Review by bayandin — APPROVED** [review]

**Comment by sharnoff** [comment]
(maybe off-topic, but why is autoscaling review requested?)

**Comment by Bodobolero** [comment]
> (maybe off-topic, but why is autoscaling review requested?)

good question, I did not explicitly request it, maybe some code ownership rules of affected files?

**Pull request merged**